### PR TITLE
Add DDS Replay application

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
After:
 - https://github.com/eProsima/DDS-Pipe/pull/39
 - https://github.com/eProsima/dev-utils/pull/62

### CODE REVIEW

File: docs/rst/replaying/usage/configuration.rst [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+.. _replayer_topic_filtering:
+
+Topic Filtering
+^^^^^^^^^^^^^^^
+
+As seen in :ref:`DDS Recorder topic filtering <recorder_topic_filtering>`, a user can define a set of rules to only record DDS :term:`Topics<Topic>` of interest.
+
+In addition to the filters applied to |ddsrecorder| when recording, |ddsreplayer| also allows filtering of DDS :term:`Topics<Topic>`.
+That is, it allows to define the DDS Topics' data that is going to be replayed by the application.
```
Comment (irenebm) [inline docs/rst/replaying/usage/configuration.rst:43]:
```suggestion
That is, it allows to define the DDS Topics data that is going to be replayed by the application.
```

File: docs/rst/replaying/usage/usage.rst [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+|eddsrecord| depends on ``fastrtps``, ``fastcdr`` and ``ddspipe`` libraries.
+In order to correctly execute the replayer, make sure that ``fastrtps``, ``fastcdr`` and ``ddspipe`` are properly sourced.
+
+.. code-block:: bash
+
+    source <path-to-fastdds-installation>/install/setup.bash
+    source <path-to-ddspipe-installation>/install/setup.bash
+    source <path-to-ddsrecorder-installation>/install/setup.bash
```
Comment (irenebm) [inline docs/rst/replaying/usage/usage.rst:47]:
```suggestion
    source <path-to-ddsreplayer-installation>/install/setup.bash
```

File: ddsreplayer/src/cpp/user_interface/arguments_configuration.cpp [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        {
+            print_version();
+            return ProcessReturnCode::version_argument;
+        }
+
+        // if (options[optionIndex::INPUT_FILE].count() == 0)
+        // {
+        //     logError(DDSREPLAYER_ARGS, "ERROR: Input file argument (-i | --input-file) is required." );
+        //     option::printUsage(fwrite, stdout, usage, columns);
+        //     return ProcessReturnCode::incorrect_argument;
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsreplayer/src/cpp/user_interface/arguments_configuration.cpp:245]: This is zombie code, it used to be mandatory when only provided through CLI (now also YAML).

File: ddsreplayer/src/cpp/user_interface/arguments_configuration.cpp [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+            return ProcessReturnCode::help_argument;
+        }
+
+        if (options[optionIndex::VERSION])
+        {
+            print_version();
+            return ProcessReturnCode::version_argument;
+        }
+
+        // if (options[optionIndex::INPUT_FILE].count() == 0)
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsreplayer/src/cpp/user_interface/arguments_configuration.cpp:241]: See comment above.

File: resources/configurations/replayer/allowlist_config.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    - name: "*"
+
+  # Blocked topics
+  blocklist:
+    - name: "add_blocked_topics_list_here"
+
+# Replayer configuration
+replayer:
+  # Input file settings
+  input-file: input_file.mcap
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply resources/configurations/replayer/allowlist_config.yaml:17]: We agreed offline they are not exactly comparable and to keep this as it is.

File: ddsrecorder_participants/include/ddsrecorder_participants/replayer/McapReaderParticipantConfiguration.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace eprosima {
+namespace ddsrecorder {
+namespace participants {
+
+/**
+ * TODO comment
+ */
+struct McapReaderParticipantConfiguration : ddspipe::participants::ParticipantConfiguration
+{
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrecorder_participants/include/ddsrecorder_participants/replayer/McapReaderParticipantConfiguration.hpp:31]: Input file is a constructor argument, so no need to include it here again.

File: ddsrecorder_participants/src/cpp/replayer/ReplayerParticipant.cpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        discovery_database)
+{
+    // Delete endpoint discovery/removal callbacks inserted in DDS-Pipe core.
+    // This is to avoid the creation of useless bridges and tracks created when discovering endpoints in topics other
+    // than the ones present in MCAP.
+    discovery_database_->clear_all_callbacks();
+}
+
+std::shared_ptr<IReader> ReplayerParticipant::create_reader(
+        const ITopic& /* topic */)
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrecorder_participants/src/cpp/replayer/ReplayerParticipant.cpp:43]: This is an IParticipant overriden method, cannot do anything about it.

File: ddsrecorder_participants/src/cpp/replayer/McapReaderParticipant.cpp [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    return false;
+}
+
+bool McapReaderParticipant::is_rtps_kind() const noexcept
+{
+    return false;
+}
+
+std::shared_ptr<IWriter> McapReaderParticipant::create_writer(
+        const ITopic& /* topic */)
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply ddsrecorder_participants/src/cpp/replayer/McapReaderParticipant.cpp:72]: Same as above.

File: docs/rst/appendixes/glossary.rst [hunk 9]
```diff
@@ -6,7 +6,80 @@
...
+DDS nomenclature
+================
+
+.. glossary::
+
+    DataReader
+        DDS element that subscribes to a specific Topic.
+        It belong to one and only one Participant, and it is uniquely identified by a Guid.
+
+        See `Fast DDS documentation <https://fast-dds.docs.eprosima.com/en/v2.4.1/fastdds/dds_layer/subscriber/subscriber.html>`__
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/appendixes/glossary.rst:34]: Why? Actually it should be a reference to DataReader, I changed that.

[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/appendixes/glossary.rst:34]: I also noted Fast-DDS documentation version is pretty old, please do not resolve this conversation until we solve it.

File: docs/rst/installation/docker.rst [hunk 10]
```diff
@@ -27,15 +27,15 @@ The steps to run |ddsrecorder| in a Docker container are explained below.
...
     |br|
 
-#.  Build a |ddsrecorder| configuration YAML file on the local machine.
-    This will be the |ddsrecorder| configuration file that runs inside the Docker container.
+#.  Build a |ddsrecord| configuration YAML file on the local machine.
+    This will be the |ddsrecord| configuration file that runs inside the Docker container.
     To continue this installation manual, let's use one of the configuration files provided in the
-    :ref:`Tutorials <tutorials_basic_example>` section.
-    Open your preferred text editor and copy a full example from the :ref:`Tutorials <tutorials_basic_example>` section
+    :ref:`Tutorials <recorder_tutorials_basic_example>` section.
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/installation/docker.rst:37]: I see one in the Foxglove tutorial.

File: docs/rst/recording/usage/configuration.rst [hunk 11]
```diff
@@ -301,13 +305,13 @@ If ``max-reception-rate`` is also set, downsampling applies to messages that alr
...
 
-.. _usage_configuration_remote_controller:
+.. _recorder_usage_configuration_remote_controller:
 
 Remote Controller
 -----------------
 
 Configuration of the DDS remote control system.
-Please refer to :ref:`Remote Control <remote_control>` for further information on how to use |ddsrecorder| remotely.
+Please refer to :ref:`Remote Control <recorder_remote_control>` for further information on how to use |ddsrecorder| remotely.
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/recording/usage/configuration.rst:314]: We break when reaching a full stop.

File: docs/rst/replaying/getting_started/usage_example.rst [hunk 12]
```diff
@@ -0,0 +1,73 @@
+.. include:: ../../exports/alias.include
+
+.. _replayer_getting_started_usage_example:
+
+################
+Example of usage
+################
+
+This example will serve as a hands-on tutorial, aimed at introducing some of the key concepts and features that
+|eddsrecord| replay application (|ddsreplayer| or ``ddsreplayer``) has to offer.
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/replaying/getting_started/usage_example.rst:10]: I would stay consistent.

File: docs/rst/appendixes/glossary.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/appendixes/glossary.rst]: Updated all Fast-DDS documentation references to version 2.10.1

File: docs/rst/developer_manual/installation/configuration/cmake_options.rst [hunk 13]
```diff
@@ -26,7 +26,7 @@ These options allow the developer to enable/disable certain |eddsrecord| setting
           ``Debug``
         - ``Release``
     *   - :class:`BUILD_DDSRECORDER_CONTROLLER`
-        - Build the |eddsrecord| remote |br| controller application. |br|
+        - Build the |ddsrecorder| remote |br| controller application. |br|
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/developer_manual/installation/configuration/cmake_options.rst:29]: Breaking lines won the referendum.

File: docs/rst/recording/usage/configuration.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/recording/usage/configuration.rst:314]: Broke line.

**Review by irenebm — APPROVED** [review]
LGTM

**Review by irenebm — APPROVED** [review]
LGTM

**Pull request merged**